quil-py/src/program/mod.rs

```diff
@@ -218,8 +225,75 @@ impl PyProgram {
             .to_owned())
     }
 
-    pub fn __str__(&self) -> String {
-        self.as_inner().to_string()
+    pub fn resolve_placeholders(&mut self) {
+        self.as_inner_mut().resolve_placeholders();
+    }
+
+    // TODO: Pre-resolve placeholders into a map so a non-panic error can be raised?
```

Comment:
If the given function raises an exception or returns an incompatible type, we can't bubble up an error from the resolver function, which is why the current implementation uses `assert!`. Because these are pyo3 bindings, that isn't as bad as it would be in a pure Rust context. `pyo3` wraps Rust panics in a `PanicException`, that gets raised like any other exception. The context string from the `assert!` gets included in the exception too, so the error is actually pretty clear if you can get past the `PanicException`.

Alternatively, we can use the given function to populate a pair of `HashMap` for the placeholders and use the resulting `HashMap`s to build the resolver functions. If populating the `HashMap`s fail for any reason we can bubble up the error in the usual `PyResult`. This requires methods that don't exist on `Program` yet, and I'm not sure if it would make sense to make such methods public either.

Reply:
Admittedly, I don't quite understand your proposed alternative, but if the current implementation gives a useful Python exception I would say that's enough.